cpp/src/barretenberg/honk/proof_system/prover_library.cpp

```diff
@@ -0,0 +1,360 @@
+
+    for (size_t i = 0; i < circuit_size; ++i) {
+
+        // Compute i'th element of f via Horner (see definition of f above)
+        T0 = lookup_index_selector[i];
+        T0 *= eta;
+        T0 += wires[2][(i + 1) & block_mask] * column_3_step_size[i];
+        T0 += wires[2][i];
+        T0 *= eta;
+        T0 += wires[1][(i + 1) & block_mask] * column_2_step_size[i];
+        T0 += wires[1][i];
+        T0 *= eta;
+        T0 += wires[0][(i + 1) & block_mask] * column_1_step_size[i];
+        T0 += wires[0][i];
+        T0 *= lookup_selector[i];
+
+        // Set i'th element of polynomial q_lookup*f + γ
+        accumulators[0][i] = T0;
+        accumulators[0][i] += gamma;
+
+        // Compute i'th element of t via Horner
+        T0 = tables[3][(i + 1) & block_mask];
+        T0 *= eta;
+        T0 += tables[2][(i + 1) & block_mask];
+        T0 *= eta;
+        T0 += tables[1][(i + 1) & block_mask];
+        T0 *= eta;
+        T0 += tables[0][(i + 1) & block_mask];
+
+        // Set i'th element of polynomial (t + βt(Xω) + γ(1 + β))
+        accumulators[1][i] = T0 * beta + next_table;
+        next_table = T0;
+        accumulators[1][i] += gamma_beta_constant;
+
+        // Set value of this accumulator to (1 + β)
+        accumulators[2][i] = beta_constant;
+
+        // Set i'th element of polynomial (s + βs(Xω) + γ(1 + β))
+        accumulators[3][i] = sorted_list_accumulator[(i + 1) & block_mask];
+        accumulators[3][i] *= beta;
+        accumulators[3][i] += sorted_list_accumulator[i];
+        accumulators[3][i] += gamma_beta_constant;
+    }
+
+    // Step (2)
+
+    // Note: This is a small multithreading bottleneck, as we have only 4 parallelizable processes.
+    for (auto& accum : accumulators) {
+        for (size_t i = 0; i < circuit_size - 1; ++i) {
+            accum[i + 1] *= accum[i];
+        }
+    }
+
+    // Step (3)
+
+    // Compute <Z_lookup numerator> * ∏_{j<i}∏_{k<j}S_k
+    Fr inversion_accumulator = Fr::one();
+    for (size_t i = 0; i < circuit_size - 1; ++i) {
+        accumulators[0][i] *= accumulators[2][i];
+        accumulators[0][i] *= accumulators[1][i];
+        accumulators[0][i] *= inversion_accumulator;
+        inversion_accumulator *= accumulators[3][i];
+    }
+    inversion_accumulator = inversion_accumulator.invert(); // invert
+
+    // Compute [Z_lookup numerator] * ∏_{j<i}∏_{k<j}S_k / ∏_{j<i+1}∏_{k<j}S_k = <Z_lookup numerator> /
+    // ∏_{k<i}S_k
+    for (size_t i = circuit_size - 2; i != std::numeric_limits<size_t>::max(); --i) {
+        // N.B. accumulators[0][i] = z_lookup[i + 1]
+        accumulators[0][i] *= inversion_accumulator;
+        inversion_accumulator *= accumulators[3][i];
+    }
+
+    Polynomial z_lookup(key->circuit_size);
+    // Initialize 0th coefficient to 0 to ensure z_perm is left-shiftable via division by X in gemini
+    z_lookup[0] = Fr::zero();
```

Comment:
I was not planning to make these functions shareable between plonk and honk, at least not for now. That would involve mucking around in the widgets (where the corresponding plonk versions live). But it's a good point and the logic is very similar so I'll make an issue to highlight that this could be done in the future if desired.